Database-level access control

dvkashapov opened the pull request:
## API changes and user behavior:

- [x] Default behavior for database access.

Default is `alldbs` permissions.


### Database Permissions (`db=`)
- [x] Accessing particular database

```
> ACL SETUSER test1 on +@all ~* resetdbs db=0,1 nopass
"user test1 on nopass sanitize-payload ~* resetchannels db=0,1 +@all"
```

- [x] (Same behavior without usage of `resetdbs`)
```
> ACL SETUSER test1 on +@all ~* db=0,1 nopass
"user test1 on nopass sanitize-payload ~* resetchannels db=0,1 +@all"
```

- [x] Multiple selector can be provided
```
> ACL SETUSER test1 on nopass (db=0,1 +@write +select ~*) (db=2,3 +@read +select ~*)
"user test1 on nopass sanitize-payload resetchannels alldbs -@all (~* resetchannels db=0,1 -@all +@write +select) (~* resetchannels db=2,3 -@all +@read +select)"
```

- [x] Restricting special commands which access databases as part of the command. 

The user needs to have access to both the commands and db(s) part of the command to run these commands.

1. SWAPDB
2. SELECT
3. MOVE - (Select command would have went through for the source database). Have access for the target database.
4. COPY


- [x] Restricting special commands which doesn't specify database number, however, accesses multiple databases.

The user needs to have access to both the commands and all databases (`alldbs`) to run these commands.

1. FLUSHALL - Access all databases
2. CLUSTER commands that access all databases:
    - CANCELSLOTMIGRATIONS
    - MIGRATESLOTS

- [x] New connection establishment behavior
New client connection gets established to DB 0 by default. Authentication and authorisation are decoupled and the user can connect/authenticate and further perform `SELECT` or other operation that do not access keyspace.

(Do we want to extend HELLO?) Alternative suggestion by @madolson: Extend `HELLO` command to pass the dbid to which the user should get connected after authentication if they have right set of permission. I think it will become a long poll for adoption.

- [x] Observability
Extend `ACL LOG` to log user which received denied permission error while accessing a database.

- [x] Module API
* Introduce module API `int VM_ACLCheckPermissions(ValkeyModuleUser *user, ValkeyModuleString **argv, int argc, int dbid, ValkeyModuleACLLogEntryReason *denial_reason);` 
* Stop support of `VM_ACLCheckCommandPermissions()`.

Resolves: #1336

dvkashapov committed:
Add database-level access control

- Implemented database permissions using `db=<id>`, `db!=<id>`, `alldbs`, `resetdbs` ACL rules
- Added SELECTOR_FLAG_ALLDBS and SELECTOR_FLAG_DBLIST_NEGATED flags
- Updated SELECT, MOVE, SWAPDB, FLUSHDB commands with CMD_CROSS_DB flag, FLUSHALL - CMD_ALL_DBS
- Extended ACL checks with database access verification
- Added ACL_DENIED_DB error type for database permission violations
- Maintained backward compatibility with default access to all databases

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

madolson commented on src/acl.c:
```diff
@@ -997,6 +1097,15 @@ static aclSelector *aclCreateSelectorFromOpSet(const char *opset, size_t opsetle
+ * db=<dbid>    Add database ID to the set of allowed IDs, any other database ID
+ *              will be disallowed. May be used with `,` for assingning multiple 
+ *              allowed IDs (e.g "db=1,2,3").
+ * db!=<dbid>   Add database ID to the set of disallowed IDs, any other database ID
+ *              will be disallowed. May be used with `,` for assingning multiple
+ *              disallowed IDs (e.g "db!=4,5,6").
```
We don't support negative keys, so I'm a little less clear why we need negative databases. The original design for ACLs was that you grant access on small sets of data to users. If we can remove this part of the syntax, I think that will make it easier overall to understand.

dvkashapov replied:
Makes sense, will do. Initially it was assumed that the user would specify `db!=15` and there would be no need to list the other 15 databases with `db=`, but I see that it complicates syntax.

madolson commented on src/module.c:
```diff
@@ -6527,7 +6527,10 @@ ValkeyModuleCallReply *VM_Call(ValkeyModuleCtx *ctx, const char *cmdname, const
         int acl_retval;
 
-        acl_retval = ACLCheckAllUserCommandPerm(user, c->cmd, c->argv, c->argc, &acl_errpos);
+        int dbid = (c->flag.multi) ? 
```
Additionally for modules, we'll eventually probably want:
1. A new API for checking if the user can access a DB
2. New flags for CrossDB/All DBs. Might be able to just skip this for now.

dvkashapov replied:
Both would be useful. Seems like API and flags won't be a lot of code, should we do this in this PR or do as a follow-up with another Issue?

madolson commented on src/multi.c:
```diff
@@ -97,6 +98,15 @@ void queueMultiCommand(client *c, uint64_t cmd_flags) {
     mc->argv_len = c->argv_len;
     mc->slot = c->slot;
+    
+    if (mc->cmd->proc == selectCommand && mc->argc > 1) {
```
We have some other code like this in cluster.c, we probably should write helpers for each command that accesses DBs and provide a helper function. Basically all functions can access up to two DBs, so we should be able to construct an interface around that.

dvkashapov replied:
I wrote an implementation that extends serverCommand with hepler function that extracts db args from command, is that what you meant and what we should use in that situation? Or you meant something else that could potentially also resolve same hardcoded commands in cluster.c?

madolson replied:
I think what you are saying makes sense.

dvkashapov committed:
Fix selector flag assignment and apply clang-format

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
Extend serverCommand with get_dbid_args, unify ACL checks

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

hpatro commented on src/acl.c:
```diff
@@ -1635,18 +1759,59 @@ static void cleanupACLKeyResultCache(aclKeyResultCache *cache) {
- * to access a key or channel that are not among the specified patterns. */
+ * command, the second, third and fourth if the command is denied because the selector is trying
+ * to access a key, channel or database that are not among the specified patterns. */
 static int ACLSelectorCheckCmd(aclSelector *selector,
```
FLUSHALL and FLUSHDB also needs to be handled under this code flow.

dvkashapov replied:
FLUSHALL is handled in `} else if ((cmd->flags & CMD_ALL_DBS) ...`
FLUSHDB is in `if (cmd->flags & CMD_CROSS_DB) {`
or you mean something else?

hpatro commented on src/acl.c:
```diff
@@ -1635,18 +1759,59 @@ static void cleanupACLKeyResultCache(aclKeyResultCache *cache) {
     int ret;
+
+    /* Check database level permissions based on command flags. */
+    if (cmd->flags & CMD_CROSS_DB) {
```
Do we need add this flag as well as add get dbid proc ?

If dbid proc is NULL, that would imply it's not a cross DB operation.

dvkashapov replied:
The main problem here is that someone can easily forget to add implementation of dbid proc for new cross db command. Your logic is valid but I think we have to handle that somehow, what do you think?

hpatro replied:
At this point we introduce commands almost negligibly. 

Going by your above logic, they could also not add the CROSS_DB flag.

hpatro commented on src/acl.c:
```diff
@@ -1635,18 +1759,59 @@ static void cleanupACLKeyResultCache(aclKeyResultCache *cache) {
+            if (dbids) {
+                for (int i = 0; i < count; i++) {
+                    if (!ACLSelectorCanAccessDb(selector, dbids[i])) {
+                        if (keyidxptr) {
+                            if (cmd->proc == selectCommand)
+                                *keyidxptr = 1;
+                            else if (cmd->proc == moveCommand)
+                                *keyidxptr = 2;
+                            else if (cmd->proc == swapdbCommand)
+                                *keyidxptr = (i == 0) ? 1 : 2;
+                            else
+                                *keyidxptr = 0;
+                        }
```
need to handle flushdb here. This part of the code doesn't seem very readable. Let me think more and come up with something.

dvkashapov replied:
Agreed that this needs improvement, don't really have an idea for how to do it cleaner.

dvkashapov replied:
> need to handle flushdb here

It is handled as a general case in last else if because FLUSHDB is related to current dbid

> This part of the code doesn't seem very readable

Added todo here, still can't think of clean way to set keyidxptr

hpatro commented on src/acl.c:
```diff
@@ -935,6 +959,70 @@ static void ACLAddAllowedFirstArg(aclSelector *selector, unsigned long id, const
+    }
+
+    // Reject empty list, trailing commas or more than 1 consecutive commas
+    if (strlen(dblist) == 0 || dblist[0] == ',' ||
+        dblist[strlen(dblist) - 1] == ',' || strstr(dblist, ",,")) {
+        zfree(dblist);
+        intsetFree(new_dbs);
+        errno = EINVAL;
+        return C_ERR;
+    }
```
All the other resources, only allow one token per symbol, we should stick with that behavior.

dvkashapov replied:
> one token per symbol

Seems aligned with ACL syntax, will do.

hpatro commented on src/acl.c:
```diff
@@ -1011,6 +1105,21 @@ static int ACLSetSelector(aclSelector *selector, const char *op, size_t oplen) {
+            intsetFree(selector->dbs);
+            selector->dbs = intsetNew();
+        }
+    } else if (strncmp(op, "db=", 3) == 0 || strncmp(op, "DB=", 3) == 0) {
```
you could do a case insensitive comparison `strncasecmp(op, "db=", 3)`.

hpatro commented on src/acl.c:
```diff
@@ -1635,18 +1759,59 @@ static void cleanupACLKeyResultCache(aclKeyResultCache *cache) {
+                zfree(dbids);
+            }
+        } else {
+            /* Command has flag CMD_CROSS_DB but has no function to get its db id args */
+            return ACL_NOT_IMPLEMENTED;
```
This should never happen.

dvkashapov replied:
If we already decided on removing `CMD_CROSS_DB` flag and only looking at `get dbid proc` presence, than this is not needed, will remove.

hpatro commented on src/acl.c:
```diff
@@ -2709,6 +2880,21 @@ sds getAclErrorMessage(int acl_res, user *user, struct serverCommand *cmd, sds e
+        } else {
+            return sdsnew("No permissions to access database");
+        }
+    case ACL_NOT_IMPLEMENTED:
+        return sdscatfmt(sdsempty(),
+                         "Command '%S' has flag CMD_CROSS_DB "
+                         "but has no function to get its database id args, "
+                         "see serverCommand in server.h",
+                         cmd->fullname);
```
Remove.

dvkashapov replied:
Will do.

hpatro reviewed (commented):
Thanks for working on this @JoBeR007. 

High level comment:

* I would prefer us to come up with a symbol for DB, let's say `^` and use that instead of `db=` as prefix to have the same experience as other Then it would look like,

All database access: `ACL SETUSER alice ^*`
Selective database access: `ACL SETUSER bob ^1 ^2`

* With DB I find providing both ALLOW and DENY would be helpful. Possibly extend it to other resources in a separate PR.
  
Restrict admin db access via (`-^?`):

Allow all except db 0: `ACL SETUSER alice -^0`

* For `FLUSHALL` and `FLUSHDB` the command can get accepted as ASYNC and while it's getting executed the permission could change. I think the correct behavior is to still process the command completely even if the permission has changed at a later point. (The PR has the correct behavior).

dvkashapov commented:
@madolson @hpatro Hello! It would be awesome to merge this in 9.1!
I think I will rewrite to `ACL SETUSER alice ^dbid` this week. There's some talks about named databases and in that case something like `ACL SETUSER alice ^db_name1 ^db_name2` looks better than `ACL SETUSER bob db=db_name1,db_name2`, what do you think?

dvkashapov commented:
> still per-selector

But then db+=1,2,3 would only apply to root selector, and will not be useful for user with >1 selector, same for db-=1,2,3 case

dvkashapov committed:
Run generate commands

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
Merge remote-tracking branch 'upstream/unstable' into unstable

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
Fix tests and apply clang

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
Delete CROSS_DB and NOT_IMPLEMENTED

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
Delete CROSS_DB from commands.def

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
Add db+= and db-= syntax

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
Fix comment style

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
clang-format fix

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov commented:
I've also added `db-=` syntax but in order to handle selector with alldbs I had to add part where we convert from alldbs to explicit list of dbs, what do you think about this part?

dvkashapov committed:
Refactor and add more tests

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov commented:
Hello @hpatro @madolson! I think PR is ready for proper review, the only things left are module API/flags and decide on some of topics raised earlier:
1. Should we make an assumption that all `@admin` commands need alldbs access?
2. Should we make FLUSHALL flush all **available to that user** dbs? (I think we should not do this, would be too big of a breaking change)
3. Decide on secure by default.

dvkashapov committed:
Merge remote-tracking branch 'upstream/unstable' into unstable

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

hpatro commented on src/acl.c:
```diff
@@ -934,6 +958,93 @@ static void ACLAddAllowedFirstArg(aclSelector *selector, unsigned long id, const
+    /* For removal operations with ALLDBS flag, convert to explicit list */
+    if (!add_dbs && (selector->flags & SELECTOR_FLAG_ALLDBS)) {
+        new_dbs = intsetNew();
+        if (!new_dbs) return NULL;
```
Is this possible ? I don't think we perform this check anywhere. We would rather crash if we aren't able to allocate memory.

dvkashapov replied:
Probably a good idea, thanks!

hpatro commented on src/acl.c:
```diff
@@ -934,6 +958,93 @@ static void ACLAddAllowedFirstArg(aclSelector *selector, unsigned long id, const
+    intset *new_dbs;
+
+    /* For removal operations with ALLDBS flag, convert to explicit list */
+    if (!add_dbs && (selector->flags & SELECTOR_FLAG_ALLDBS)) {
```
Would it be handy to maintain two list 

1. Allow list
2. Deny list

We might be able to avoid O(DBCOUNT) operation.

hpatro replied:
Alternative suggestion: https://github.com/valkey-io/valkey/pull/2309#discussion_r2532274461

dvkashapov replied:
Note: decided to not include `db-=` syntax

hpatro commented on src/acl.c:
```diff
@@ -934,6 +958,93 @@ static void ACLAddAllowedFirstArg(aclSelector *selector, unsigned long id, const
+    intset *new_dbs = ACLInitDatabaseIntset(selector, add_dbs);
+    if (!new_dbs) {
+        return ACLDatabasePermissionError(NULL, NULL, ENOMEM);
+    }
```
I think the only scenario in which you return `NULL` isn't required. So, we can get rid of this as well.


```suggestion
```

dvkashapov replied:
Yes, thanks!

hpatro commented on src/acl.c:
```diff
@@ -934,6 +958,93 @@ static void ACLAddAllowedFirstArg(aclSelector *selector, unsigned long id, const
+    char *dblist = zstrdup(dbs_str);
+    if (!dblist) {
+        return ACLDatabasePermissionError(new_dbs, NULL, ENOMEM);
+    }
```
```suggestion
```

dvkashapov replied:
Will do!

hpatro commented on src/acl.c:
```diff
@@ -934,6 +958,93 @@ static void ACLAddAllowedFirstArg(aclSelector *selector, unsigned long id, const
+        selector->flags &= ~SELECTOR_FLAG_ALLDBS;
+    }
+
+    char *dblist = zstrdup(dbs_str);
```
Would suggest creating a `sds` and operate on it. improves `strlen` perf and more.

dvkashapov replied:
Rewrote to sds usage, PTAL

dvkashapov committed:
Fix reason in ACL LOG and add test

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
Use cmd->fullname when no argpos

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
Add module api for db-level check and test

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov commented:
@hpatro added `VM_ACLCheckDbPermissions` that checks db-level permissions for user, do we need anything else?

ranshid commented on src/acl.c:
```diff
@@ -810,6 +817,23 @@ static sds ACLDescribeSelector(aclSelector *selector) {
+    /* Database permissions. */
+    if (selector->flags & SELECTOR_FLAG_ALLDBS) {
+        res = sdscatlen(res, "alldbs ", 7);
+    } else {
+        res = sdscatlen(res, "db+=", 4);
+        uint32_t len = intsetLen(selector->dbs);
+        for (uint32_t i = 0; i < len; i++) {
+            int64_t dbid;
+            if (intsetGet(selector->dbs, i, &dbid)) {
+                if (i > 0) res = sdscatlen(res, ",", 1);
+                res = sdscatfmt(res, "%I", dbid);
+            }
+        }
+
+        res = sdscatlen(res, " ", 1);
```
Probably a niche, but did you consider the case of disable access to specific dbs when the dbnum is very large?
I guess it would be nice if we can make a special handling of the case when intsetLen >> 0 and make a:`alldbs db-=...`. I guess it is also possible to make it in a followup improvement, but lets just make sure we create an issue for it (if we think it is valuable)

hpatro replied:
I had apprehension on the same code flow. My suggestion https://github.com/valkey-io/valkey/pull/2309#discussion_r2532274461 . WDYT ?

dvkashapov replied:
I did and my first proposal had negated syntax support, but we decided to not go down that route because of possible confusion for user and goal of ACL's - to grant small sets of permissions.

dvkashapov replied:
Note: decided to not include `db-=` syntax

hpatro commented on src/acl.c:
```diff
@@ -934,6 +958,93 @@ static void ACLAddAllowedFirstArg(aclSelector *selector, unsigned long id, const
+
+        for (int i = 0; i < server.dbnum; i++) {
+            uint8_t success;
+            new_dbs = intsetAdd(new_dbs, i, &success);
```
Alternative to my previous suggestion of maintaining couple of list, if we would like to stick a single list. As `server.dbnum` is immutable, we could simply maintain a static const allocation of intset with all dbs added once and perform a `intsetDup` of that for any `ACL SETUSER` operation.

dvkashapov replied:
That seems like wasted memory for nothing in most of use cases, no?

dvkashapov replied:
Note: decided to not include `db-=` syntax

hpatro commented on src/acl.c:
```diff
@@ -1011,6 +1130,24 @@ static int ACLSetSelector(aclSelector *selector, const char *op, size_t oplen) {
+            intsetFree(selector->dbs);
+            selector->dbs = intsetNew();
+        }
+    } else if (!strcasecmp(op, "resetdbs")) {
+        selector->flags |= SELECTOR_FLAG_ALLDBS;
+        if (selector->dbs) {
+            intsetFree(selector->dbs);
+            selector->dbs = intsetNew();
+        }
```
This seems incorrect. It should be no permission to any dbs.


```suggestion
    } else if (!strcasecmp(op, "resetdbs")) {
        selector->flags &= ~SELECTOR_FLAG_ALLDBS;
        if (selector->dbs) {
            intsetFree(selector->dbs);
            selector->dbs = intsetNew();
        }
```

dvkashapov replied:
Ok, got it, seems like there was some misunderstanding because that's exactly why I asked: 
> Should resetdbs right now be secure by default? (do not set alldbs flag and require explicit db list)
> No. Have added my thought above.

There's problem with secure resetdbs approach. If we default to alldbs but reset to no dbs, then in` } else if (!strcasecmp(op, "reset")) {` we have to `serverAssert(ACLSetUser(u, "alldbs", -1) == C_OK);` and not `serverAssert(ACLSetUser(u, "resetdbs", -1) == C_OK);`, which is unintuitive, what do you think?

Anyway, I made resetdbs  - no permission to any dbs, please take a look.

hpatro commented on src/acl.c:
Not particular for this PR, . Was it a security requirement behind introducing a non-verbose mode and hide the user/resource from the error response ?

dvkashapov replied:
I also thought that it would be useful for user to know..

madolson replied:
No, it was to avoid a breaking change at some point. The dry-run needed to give some more information, but we didn't want to change the response.

hpatro reviewed (commented):
Seems mostly good on `acl.c` barring the resetdbs behavior and code performance on each `ACL SETUSER` operation. 

Going through the remaining files.

hpatro commented:
> @hpatro added `VM_ACLCheckDbPermissions` that checks db-level permissions for user, do we need anything else? Also updated top comment.

Why would one use this check independently ?

> Question: We need to handle db-level checks in `VM_ACLCheckCommandPermissions` because it calls `ACLCheckAllUserCommandPerm` which needs dbid, what should we do in this scenario?

There are two alternatives to deal with this. 

1. Stop supporting the API with 9.1. We could alter the API or recommend module devs to use a new API `VM_ACLCheckPermissions` which takes in all the parameters required. 
2. Structure our code in a way that still supports the existing API but the command isn't guaranteed to succeed if ACL database restriction feature is used. 

I would recommend going with option 1. @allenss-amazon thoughts?

hpatro commented on src/acl.c:
```diff
@@ -1011,6 +1130,24 @@ static int ACLSetSelector(aclSelector *selector, const char *op, size_t oplen) {
+        if (result != C_OK) return result;
+    } else if (strncasecmp(op, "db-=", 4) == 0) {
+        int result = ACLSetSelectorDatabasePermissions(selector, op + 4, 0);
+        if (result != C_OK) return result;
```
Adding and removing single dbid leads to incorrect response

```
127.0.0.1:6379> ACL LIST
1) "user default on nopass sanitize-payload ~* &* alldbs +@all"
2) "user hp on nopass sanitize-payload resetchannels db+=1 -@all"
127.0.0.1:6379> ACL SETUSER hp db-=1
OK
127.0.0.1:6379> ACL LIST
1) "user default on nopass sanitize-payload ~* &* alldbs +@all"
2) "user hp on nopass sanitize-payload resetchannels db+= -@all"
127.0.0.1:6379> ACL SETUSER dk nopass
OK
127.0.0.1:6379> ACL LIST
1) "user default on nopass sanitize-payload ~* &* alldbs +@all"
2) "user dk off nopass sanitize-payload resetchannels alldbs -@all"
3) "user hp on nopass sanitize-payload resetchannels db+= -@all"
```

`dk` and `hp` user should have the same response.

dvkashapov replied:
But why? Shouldn't `hp` have no allowed dbs? Current behaviour:
```
127.0.0.1:6379> acl setuser hp on nopass sanitize-payload resetchannels db+=1 -@all
OK
127.0.0.1:6379> ACL SETUSER hp db-=1
OK
127.0.0.1:6379> ACL LIST
1) "user default on nopass sanitize-payload ~* &* alldbs +@all"
2) "user hp on nopass sanitize-payload resetchannels resetdbs -@all"
127.0.0.1:6379>
```

hpatro replied:
You're right. I was confused with empty allow list db+=. Was there some issue with it?

dvkashapov replied:
Yes, resetdbs was printed as `db+=`, fixed it

hpatro commented on src/acl.c:
```diff
@@ -934,6 +958,93 @@ static void ACLAddAllowedFirstArg(aclSelector *selector, unsigned long id, const
+        if (errno == ERANGE || endptr == token || *endptr != '\0' ||
+            dbid < 0 || dbid >= server.dbnum) {
+            return ACLDatabasePermissionError(new_dbs, dblist, EINVAL);
+        }
```
Would be better for user to understand what's incorrect in the command. WDYT ?

Current response is generic syntax error warning.
```
127.0.0.1:6379> config get databases
1) "databases"
2) "16"
127.0.0.1:6379> acl setuser hp db+=16
(error) ERR Error in ACL SETUSER modifier 'db+=16': Syntax error
127.0.0.1:6379> acl setuser hp db+=0
OK
```

dvkashapov replied:
True, will add new errno in this case.

dvkashapov replied:
Added new `errmsg = "The provided database ID is out of range or non-numeric";`

hpatro commented on src/acl.c:
```diff
@@ -997,6 +1108,14 @@ static aclSelector *aclCreateSelectorFromOpSet(const char *opset, size_t opsetle
+ *              with `,` for adding multiple IDs (e.g "db+=1,2,3"). To override all
+ *              existing databases, use 'resetdbs' first, then 'db+='.
+ * db-=<dbid>   Remove database ID(s) from the set of allowed database IDs. May be used
+ *              with `,` for removing multiple IDs (e.g "db-=1,2,3").
```
Providing `db+=` and `db-=` in the same command doesn't seem to work correctly. In this case, user hp should have access to alldbs except 0,1.


```
127.0.0.1:6379[1]> acl setuser hp alldbs
OK
127.0.0.1:6379[1]> acl setuser hp db+=1 db-=1,0
OK
127.0.0.1:6379[1]> acl list
1) "user default on nopass sanitize-payload ~* &* alldbs +@all"
2) "user hp on nopass sanitize-payload resetchannels &test* db+= -@all +select"
```

It works fine if I skip the `db+=`.

```
127.0.0.1:6379[1]> acl setuser hp alldbs
OK
127.0.0.1:6379[1]> acl list
1) "user default on nopass sanitize-payload ~* &* alldbs +@all"
2) "user hp on nopass sanitize-payload resetchannels &test* alldbs -@all +select"
127.0.0.1:6379[1]> acl setuser hp alldbs db-=0,1
OK
127.0.0.1:6379[1]> acl list
1) "user default on nopass sanitize-payload ~* &* alldbs +@all"
2) "user hp on nopass sanitize-payload resetchannels &test* db+=2,3,4,5,6,7,8,9,10,11,12,13,14,15 -@all +select"
```

hpatro replied:
I think this is correct behavior. I was confused with the syntax.

hpatro reviewed (commented):
Seems dangerous to set databases config to large value.  The output buffer could be quite large and server processing time with setting `acl setuser` is also quite long with such high database count. This could be accounted as denial of service I believe. 

```
% src/valkey-server --daemonize yes --logfile valkey.log --databases 100000000
% src/valkey-cli
127.0.0.1:6379> acl setuser hp on resetdbs db-=0
OK
(12.23s)
127.0.0.1:6379> acl list
<response too big and takes too long to get printed out>
```

dvkashapov committed:
apply pr suggestions

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov reviewed (commented):
@hpatro Applied some of review suggestions, will get to work on the rest of them, PTAL at my comments.

dvkashapov committed:
apply clang-format

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov commented:
> There are two alternatives to deal with this.
> 
>1. Stop supporting the API with 9.1. We could alter the API or recommend module devs to use a new API >VM_ACLCheckPermissions which takes in all the parameters required.
>2. Structure our code in a way that still supports the existing API but the command isn't guaranteed to succeed if ACL >database restriction feature is used.
>I would recommend going with option 1. @allenss-amazon thoughts?

I'm OK with first way, but I guess this is another major decision and we need feedback from community on that one.
I suggest that for now we go with 2nd way and do a follow-up issue+PR that transitions to 1st API.

dvkashapov committed:
add ALL_DBS flag to migration/slot commands

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
db= syntax, acl getuser support

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
Return ACL_DENIED_DB only for R/W commands in forbidden db

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
Use sds

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
Merge remote-tracking branch 'upstream/unstable' into unstable

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
Merge remote-tracking branch 'upstream/unstable' into unstable

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
Merge remote-tracking branch 'upstream/unstable' into unstable

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
Add new errmsg for invalid dbid

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
Delete prev module api for databases

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
Add new API

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
Reorder transaction_db_id in select and more MULTI/EXEC tests

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

hpatro commented on src/acl.c:
```diff
@@ -937,6 +963,62 @@ static void ACLAddAllowedFirstArg(aclSelector *selector, unsigned long id, const
+        int64_t dbid = strtoll(tokens[i], &endptr, 10);
+
+        if (errno == ERANGE || endptr == tokens[i] || *endptr != '\0' ||
+            dbid < 0 || dbid >= server.dbnum) {
```
Wouldn't `endptr` not being NULL is sufficient to error out ?

```suggestion
        if (errno == ERANGE || *endptr != '\0' ||
            dbid < 0 || dbid >= server.dbnum) {
```

hpatro replied:
I think we should split into two cases 

1. ERANGE (out of valid database range)
2. EINVAL (encompass invalid input)

WDYT?

dvkashapov replied:
Good points, did that

hpatro commented on src/acl.c:
```diff
@@ -937,6 +963,62 @@ static void ACLAddAllowedFirstArg(aclSelector *selector, unsigned long id, const
+
+    /* Reject empty list, trailing commas or more than 1 consecutive commas */
+    if (len == 0 || dblist[0] == ',' ||
+        dblist[len - 1] == ',' || strstr(dblist, ",,")) {
```
We could avoid the `strstr` check here and add a check for empty string in the loop below (it could possibly be already handled by `strtoll`) to avoid the expensive `strstr` invocation here.

dvkashapov replied:
True, added check for 0 length in loop

hpatro commented on src/acl.c:
```diff
@@ -1647,6 +1753,19 @@ static int ACLCheckChannelAgainstList(list *reference, const char *channel, int
+
+    int found = intsetFind(selector->dbs, (int64_t)dbid);
+
+    return found;
```
nit:

```suggestion
    return intsetFind(selector->dbs, (int64_t)dbid);
```

dvkashapov replied:
Thanks!

hpatro commented on src/acl.c:
```diff
@@ -1647,6 +1753,19 @@ static int ACLCheckChannelAgainstList(list *reference, const char *channel, int
+    if (dbid < 0 || dbid >= server.dbnum || !selector->dbs)
+        return 0;
+
+    int found = intsetFind(selector->dbs, (int64_t)dbid);
```
Does this need explicit type cast? Could you update the method signature otherwise ?

dvkashapov replied:
> Does this need explicit type cast?

I think yes, intset expects int64_t but everywhere else for dbid's we use int/long long. Also I believe that cast from long long to int64_t would be resolved at compile time, so no performance penalty. (But still we would have to sign extend int dbid before ACLSelectorCanAccessDb)

dvkashapov replied:
Will drop explicit cast

hpatro commented on src/acl.c:
```diff
@@ -1666,18 +1785,55 @@ static void cleanupACLKeyResultCache(aclKeyResultCache *cache) {
+        if (dbids) {
+            for (int i = 0; i < count; i++) {
+                if (!ACLSelectorCanAccessDb(selector, dbids[i])) {
+                    /* TODO: somehow move from hardcoded cases */
```
The methods `*DbIdArgs` in db.c are also hardcoded. These two could rely on the same generic mechanism if we plan to introduce a generic dbid retriever from commands. 

However, want to avoid extensive framework like keyspec for this. Thoughts on this anyone ?

zuiderkwast replied:
Yeah this is kind of a mess. If most of this is hard-coded, why add a callback in the JSON files at all? It almost seems better with only hard-coded.

zuiderkwast replied:
However, I'm OK with the current code. We can always change it later if we find a better way.

dvkashapov replied:
> Yeah this is kind of a mess. If most of this is hard-coded, why add a callback in the JSON files at all? It almost seems better with only hard-coded.

Callback is for retrieval of dbid from argv argc, but keyidxptr is just to point to value that caused deny of access. Agree that this looks bad but I wasn't able to find a way to do that in general case, SWAPDB can have errored val at multiple positions

dvkashapov replied:
we can just set keyidxptr to 0, so errored val would be command itself, would definitely look better but not that useful

hpatro commented on src/acl.c:
```diff
@@ -1666,18 +1785,55 @@ static void cleanupACLKeyResultCache(aclKeyResultCache *cache) {
+            }
+            zfree(dbids);
+        }
+    } else if ((cmd->flags & CMD_ALL_DBS) && !(selector->flags & SELECTOR_FLAG_ALLDBS)) {
+        for (int i = 0; i < server.dbnum; i++) {
+            if (!ACLSelectorCanAccessDb(selector, i)) {
+                return ACL_DENIED_DB;
+            }
+        }
```
This would be quite expensive `O(selectors * dbnum)`. 

We could check if the inset count is not equal to dbnum we could return `ACL_DENIED_DB` early.

dvkashapov replied:
Nice, thanks!

hpatro commented on src/acl.c:
```diff
@@ -1773,6 +1929,7 @@ int ACLUserCheckKeyPerm(user *u, const char *key, int keylen, int flags, bool is
     /* If there is no associated user, the connection can run anything. */
     if (u == NULL) return ACL_OK;
 
+    // TODO: Determine what to do here and in VM function
```
What's missing?

dvkashapov replied:
Nothing, I had a questing wether we need db level checks in key permissions, answer was no

hpatro reviewed (commented):
Went through `src/acl.c`. Looks pretty good. 

Following through the other changes in the next seating 😀

hpatro commented on src/module.c:
```diff
@@ -6574,7 +6574,8 @@ ValkeyModuleCallReply *VM_Call(ValkeyModuleCtx *ctx, const char *cmdname, const
         int acl_retval;
 
-        acl_retval = ACLCheckAllUserCommandPerm(user, c->cmd, c->argv, c->argc, &acl_errpos);
+        int dbid = (c->flag.multi) ? c->mstate->transaction_db_id : (c->db ? c->db->id : -1);
```
ternary operator within another ternary operator seems a bit difficult to read. Should we convert them to if/else ?

dvkashapov replied:
I checked client creation code and we do selectDb(c, 0); so there's no need for second ternary, deleted it.

hpatro commented on src/module.c:
```diff
@@ -10156,6 +10157,52 @@ int VM_ACLCheckChannelPermissions(ValkeyModuleUser *user, ValkeyModuleString *ch
+ * On success VALKEYMODULE_OK is returned, otherwise VALKEYMODULE_ERR is returned and
+ * errno is set to one of the following values:
+ *
+ * * EINVAL: Invalid arguments (e.g., negative dbid or dbid >= server.dbnum)
```
For the explanation here, you return `ERANGE` instead of `EINVAL` in `src/acl.c` while setting a selector.

dvkashapov replied:
I'm not sure I follow, in VM_ACLCheckPermissions we set errno to EINVAL `if (dbid < 0 || dbid >= server.dbnum) {`, should we have same errno in module.c and in acl.c?

dvkashapov replied:
So the issue here is errno EINVAL for `if (dbid < 0 || dbid >= server.dbnum) {` but we need to make decision on different errno's for module API

zuiderkwast replied:
I believe EINVAL is correct if it's an invalid argument. ERANGE is used if the result can't be represented in the type. According to errno(3) man page:

    ERANGE          Result too large (POSIX.1, C99).

The module API is a public API so it's more important than acl.c, but why not use EINVAL in acl.c too? Do we need to distinguish the different cases there?

dvkashapov replied:
Yes, EINVAL is used internally as general syntax errno and I used ERANGE if provided number is out of valid range. I think acl.c usage of errno is at least unintuitive, but that’s a discussion for another time.

hpatro replied:
I would like us to stick with the same errno value across `acl.c` and `module.c`.

hpatro commented on src/module.c:
```diff
@@ -10156,6 +10157,52 @@ int VM_ACLCheckChannelPermissions(ValkeyModuleUser *user, ValkeyModuleString *ch
+ * * ENOENT: Specified command does not exist
+ * * EACCES: User does not have permission to execute the command
+ * * EPERM: User does not have permission to access a key
+ * * ECHILD: User does not have permission to access a channel
```
This errno `ECHILD` seems unrelated to channel. Should we rather return ACL_DENIED_* as method return instead of generic OK/ERR ?

dvkashapov replied:
I thought about that too and I'm all for ACL_DENIED_* but seems that in modules the convention is OK/ERR + errno.

zuiderkwast replied:
I like the suggestion. We do set `errno` in some module APIs but I think it's cleaner to return a value instead when we can. We could use `VALKEYMODULE_ACL_LOG_CHANNEL` etc. though they are meant for logging.

dvkashapov replied:
I think it's okay to reuse VALKEYMODULE_ACL_LOG_* because then it would be easy for module to log that check, just pass the retcode itself

dvkashapov replied:
`VALKEYMODULE_ERR` is defined as 1, so does `VALKEYMODULE_ACL_LOG_CMD` get assigned value of 1 in 
```
typedef enum {
    VALKEYMODULE_ACL_LOG_AUTH = 0, /* Authentication failure */
    VALKEYMODULE_ACL_LOG_CMD,      /* Command authorization failure */
    VALKEYMODULE_ACL_LOG_KEY,      /* Key authorization failure */
    VALKEYMODULE_ACL_LOG_CHANNEL,  /* Channel authorization failure */
    VALKEYMODULE_ACL_LOG_DB        /* Database authorization failure */
} ValkeyModuleACLLogEntryReason; 
```

Not sure what's the best way do deal with this, maybe negate ACL reasons or offset by +2 so that 0 and 1 are for generic OK/ERR, but that definitely is confusing for user

zuiderkwast replied:
We can't change the values of these because an old module can be used with a newer valkey and vice versa.

We can't mix them. If we return these codes for errors, we need to use zero for success. Let's compare to other module API functions to see if it makes sense or if it sticks out.

dvkashapov replied:
I meant to do something like this without changing enum values
```
switch (acl_retval) {
  case ACL_DENIED_CMD: return VALKEYMODULE_ACL_LOG_CMD + 2;
  case ACL_DENIED_KEY: return VALKEYMODULE_ACL_LOG_KEY + 2;
  case ACL_DENIED_CHANNEL: return VALKEYMODULE_ACL_LOG_CHANNEL + 2;
  case ACL_DENIED_DB: return VALKEYMODULE_ACL_LOG_DB + 2;
  default: return VALKEYMODULE_ACL_LOG_CMD + 2;
}
```
In this case 0 and 1 are reserved for general OK/ERR

dvkashapov replied:
I see possible outcomes as those:
1) Pass pointer
```
int VM_ACLCheckPermissions(ValkeyModuleUser *user, 
                          ValkeyModuleString **argv, 
                          int argc, 
                          int dbid,
                          ValkeyModuleACLLogEntryReason *denial_reason)
```
We would return OK/ERR and if ERR returned then denial_reason would be set to some value from enum ValkeyModuleACLLogEntryReason

2) Simple OK/ERR without reason

3) OK/ERR + ACL deny reason mapped to some errno value

4) Array of 2 values: OK/ERR + ValkeyModuleACLLogEntryReason

zuiderkwast replied:
I've not been reviewing this PR much in the past so I'm trying to catch up. 

The reason we're introducing this API function is that `VM_ACLCheckCommandPermissions` is incomplete and we need to provide a replacement for it that takes database id, right?

So, the starting point should be to try to provide a similar syntax so it's easy to migrate from the old function. The old one doesn't return detailed reason in errno. It only sets it to `EACCES` if it's denied. The same is true for the other API functions `VM_ACLCheckKeyPermissions` and `VM_ACLCheckChannelPermissions`. Therefore, I think it would be surprising to introduce other errno values here. Let's use only `EACCES` for permission denied.

However, I also realize that we have functions like `VM_ACLAddLogEntry` for logging, but there is no way to get this reason from the ACL checking functions.

So yeah, to summarize, I think this would be a good API:

```C
int VM_ACLCheckPermissions(ValkeyModuleUser *user, 
                           ValkeyModuleString **argv, 
                           int argc, 
                           int dbid,
                           ValkeyModuleACLLogEntryReason *denial_reason);
```

* Return OK/ERR, like the other VM_ACL checking functions.
* Set errno to EACCES like the other VM_ACL checking functions.
* `denial_reason` can be optional and you can pass NULL if you don't care.

dvkashapov replied:
> The reason we're introducing this API function is that VM_ACLCheckCommandPermissions is incomplete and we need to provide a replacement for it that takes database id, right?

Yes, exactly. Good suggestion, I will implement this API, seems to cover everything we need.

zuiderkwast replied:
Great. @hpatro can you confirm if you're OK with this suggested behavior? ↑↑

hpatro replied:
I like the suggestion more than the previous approach. 

> denial_reason can be optional and you can pass NULL if you don't care.

This won't be NULL in any case on denial I believe.

dvkashapov replied:
I think Viktor meant that if you call VM_ACLCheckPermissions(user, argv, argc, dbid, NULL) then it means you don't care about reason and you will not be able to determine it.

hpatro replied:
LOL, my bad. Sounds good.

hpatro commented on src/multi.c:
```diff
@@ -69,6 +70,7 @@ void resetClientMultiState(client *c) {
     c->mstate->cmd_inv_flags = 0;
     c->mstate->argv_len_sums = 0;
     c->mstate->alloc_count = 0;
+    c->mstate->transaction_db_id = c->db->id;
```
Why initialize it while cleaning up. I think it should suffice if we should set the value while the client is entering into multi state and update it within the multi flow if things change.

dvkashapov replied:
You're right, touchWatchedKey() triggers resetClientMultiState() and this may lead to inconsistent ACL check for client that is already in multi state but before EXEC. Or am I missing something?

hpatro commented on src/multi.c:
```diff
@@ -98,6 +100,15 @@ void queueMultiCommand(client *c, uint64_t cmd_flags) {
 
+    if (mc->cmd->proc == selectCommand && mc->argc > 1) {
+        long long target_db;
+        if (getLongLongFromObject(mc->argv[1], &target_db) == C_OK) {
```
Another hardcode touchpoint for retrieving dbid.

dvkashapov replied:
Thanks, used cmd->get_dbid_args

hpatro commented on src/server.c:
```diff
@@ -2960,6 +2960,7 @@ void initServer(void) {
     server.acl_info.user_auth_failures = 0;
     server.acl_info.invalid_channel_accesses = 0;
     server.acl_info.acl_access_denied_tls_cert = 0;
+    server.acl_info.invalid_db_accesses = 0;
```
This needs to be printed out in the info section, Look at `genValkeyInfoStringACLStats`.

hpatro replied:
Also, add a test case to verify the behavior.

dvkashapov replied:
Thanks, did that!

hpatro commented on src/server.h:
```diff
@@ -3184,11 +3193,12 @@ extern user *DefaultUser;
+#define ACL_DENIED_KEY 3
+#define ACL_DENIED_AUTH 4           /* Only used for ACL LOG entries. */
+#define ACL_DENIED_CHANNEL 5        /* Only used for pub/sub commands */
+#define ACL_INVALID_TLS_CERT_AUTH 6 /* Only used for TLS Auto-authentication */
```
Formatting seems off for this section.

dvkashapov replied:
I thought the same but in this case clang-format thinks this is okay, anyway - added comments to each line and formatted them

hpatro reviewed (commented):
Pending review of tests.

dvkashapov committed:
Apply review suggestions

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
Comment for VM_ACLCheckCommandPermissions and delete reset dbid

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

hpatro commented:
@madolson / @zuiderkwast If one of you could take a pass at this. I think it's mostly ready with couple of open questions to discuss about https://github.com/valkey-io/valkey/pull/2309#issuecomment-3649287683. And we had discussed this in one of the past weekly meetings and I guess there was a majority of TSC member approving the change. Could we update  `major-decision-` tag as well ?

dvkashapov committed:
Revert reset in multi, move errno ERANGE

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
make ACLSelectorCanAccessDb static inline

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

zuiderkwast commented:
Apparently, we've made decisions about this and they're documented in the comment: https://github.com/valkey-io/valkey/pull/2309#issuecomment-3542979702

zuiderkwast commented on src/commands/select.json:
```diff
@@ -15,6 +15,7 @@
             "CONNECTION",
             "FAST"
         ],
+        "get_dbid_args": "selectDbIdArgs",
```
There is a README.md file here among the json file that documents the json format. Please describe the new field and flag in that file.

dvkashapov replied:
Good point, will do!

dvkashapov committed:
Restrict all commands with keys/keyspace, add alldbs to FLUSHSLOT

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
Eval/func tests, clean up some lines

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
New api with denial_reason ptr

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
Merge remote-tracking branch 'upstream/unstable' into unstable

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

hpatro commented:
@dvkashapov Are all the open comments now resolved ?

dvkashapov commented:
The only one left is hardcoded check with TODO. We can come up with something better later for this, WDYT?

hpatro commented:
> The only one left is hardcoded check with TODO. We can come up with something better later for this, WDYT?

Ah, right. Let's capture others thought. I think I'm fine with the current approach unless something simpler and better is proposed. 

@zuiderkwast Any thoughts on https://github.com/valkey-io/valkey/pull/2309#discussion_r2612190893 ?

hpatro commented:
@dvkashapov reply-schemas-validator is failing.

dvkashapov commented:
Yes, I'm fixing it currently

dvkashapov committed:
Small fixes
- Update reply schema for ACL GETUSER
- Set keyidxptr for migrateCommand
- Delete ALL_DBS flag from FLUSHSLOT because it's internal command

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov commented:
@hpatro JFYI I deleted `ALL_DBS` permissions from `CLUSTER FLUSHSLOT` because I thought it was available to user, but it is internal so no need for that.

Also I believe we adressed everything, lets run more tests

hpatro commented on tests/unit/moduleapi/aclcheck.tcl:
```diff
@@ -55,6 +55,7 @@ start_server {tags {"modules acl"}} {
         set e
     } {*DENIED CHANNEL*}
 
+
```
```suggestion
```

dvkashapov commented:
Seems like test-ubuntu-32bit and test-almalinux8-tls-module flap on unstable too, I assume they're not related to this PR, just to make sure did couple of ubuntu tls test runs, they pass and do not flap

zuiderkwast commented on src/commands/acl-setuser.json:
```diff
@@ -15,6 +15,10 @@
+            ],
+            [
+                "9.1.0",
+                "Added database permission patterns."
```
Patterns?

dvkashapov replied:
Not the best word, maybe rules? I thought about just "Added database permissions." but it did not feel descriptive enough

zuiderkwast replied:
Patterns usually means glob-style patterns.

I believe "rules" matches existing terminology in the ACL documentation, but I'm not very familiar with ACL in general.

dvkashapov replied:
Changed to "rules", thanks!

dvkashapov committed:
Change wording in JSON files

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

zuiderkwast reviewed (approved):
High level LGTM, but I haven't done a complete review.

@hpatro If you take a final look, I guess we're good to go.

hpatro commented on src/intset.c:
```diff
@@ -335,3 +335,18 @@ int intsetValidateIntegrity(const unsigned char *p, size_t size, int deep) {
+    if (is) zfree(is);
+}
+
+/* Duplicate an intset */
```
```suggestion
/* Deep copy an intset */
```

hpatro commented on src/acl.c:
```diff
@@ -1662,22 +1788,69 @@ static void cleanupACLKeyResultCache(aclKeyResultCache *cache) {
+        if (dbids) {
+            for (int i = 0; i < count; i++) {
+                if (!ACLSelectorCanAccessDb(selector, dbids[i])) {
+                    /* TODO: somehow move from hardcoded cases */
```
Should we remove this TODO or leave it around ?

dvkashapov replied:
Deleted it

hpatro commented on src/acl.c:
```diff
@@ -1662,22 +1788,69 @@ static void cleanupACLKeyResultCache(aclKeyResultCache *cache) {
+                            *keyidxptr = 1;
+                        else if (cmd->proc == moveCommand)
+                            *keyidxptr = 2;
+                        else if (cmd->proc == swapdbCommand)
+                            *keyidxptr = (i == 0) ? 1 : 2;
```
Observation: for swapdb we would always return the target `dbid` if the user isn't authorized for both source/target dbid.

dvkashapov replied:
I think it's okay, we're not returning incorrect error, just pointing to first that caused denial.

hpatro commented on src/acl.c:
```diff
@@ -1859,7 +2032,7 @@ int ACLCheckAllUserCommandPerm(user *u, struct serverCommand *cmd, robj **argv,
      * 1) If no selector can execute the command, return the command.
      * 2) Return the last key or channel that no selector could match. */
-    int relevant_error = ACL_DENIED_CMD;
+    int relevant_error = ACL_DENIED_DB;
```
Note: This isn't a breaking change as the feature is opt-in. Until the user doesn't starts the usage of DB level access control they shouldn't see any denied db error. With this change however we have DB id error as of the highest priority.

dvkashapov replied:
> Note: This isn't a breaking change as the feature is opt-in. Until the user doesn't starts the usage of DB level access control they shouldn't see any denied db error.

This is true.

> With this change however we have DB id error as of the highest priority.

But this is not, in `ACLCheckAllUserCommandPerm()` for all selectors that denied access we export `relevant_error` with the highest value (because of `acl_retval > relevant_error`), currently highest priority is `ACL_DENIED_CHANNEL`

So in some sense we could have made users get `ACL_DENIED_DB` more often, example with `ACL_DENIED_DB = 7`:
```
127.0.0.1:6379> ACL SETUSER db-selector on nopass (db=0,1 +@all -@read ~write*) (db=2,3 +@all -@write ~read*)
OK
127.0.0.1:6379> auth db-selector password
OK
127.0.0.1:6379> select 0
OK
127.0.0.1:6379> set writekey value
OK
127.0.0.1:6379> get writekey
(error) NOPERM No permissions to access database
```
Currently user would get `(error) NOPERM User db-selector has no permissions to run the 'get' command` which makes more sense to me, WDYT?

hpatro replied:
ACL_DENIED_DB is of lowest priority and it seems like the right behavior.

hpatro commented on src/acl.c:
I don't see the value of `local_idxptr > last_idx`  for db level access control. However, it's not harmful I guess. Point to note we always return the highest `dbid` which failed the authorization criteria ?

dvkashapov replied:
IIUC:
`acl_retval > relevant_error` - Prefer to return higher priority errors
`acl_retval == relevant_error && local_idxptr > last_idx` - When errors have the same priority, prefer the one with the higher argument index.

I updated the comment to reflect that behaviour.

hpatro reviewed (approved):
LGTM. 

@dvkashapov could you benchmark with certain db permissions added to the default user and measure the performance change for SET/GET commands? I assume it will be a minimal drop in rps.

madolson commented on src/commands/move.json:
Was also expecting something for copy command.

dvkashapov replied:
You're correct, fixed that and added test.

madolson commented on src/commands/cluster-getkeysinslot.json:
```diff
@@ -8,7 +8,8 @@
         "command_flags": [
-            "STALE"
+            "STALE",
+            "ALL_DBS"
```
getkeysinslot should just return the data for the current DB. I don't think this is needed.

dvkashapov replied:
You're right, fixed that

madolson commented on src/commands/README.md:
```diff
@@ -93,6 +94,9 @@ following keys. To be safe, assume all of them are optional.
   * `"STREAM"`
   * `"STRING"`
   * `"TRANSACTION"`
+* `"get_dbid_args"`: The name of the C function in Valkey's source code
```
I don't think clients have any need for this, so we probably don't need to return it.

dvkashapov replied:
@zuiderkwast suggested to describe new field in commands README, looks like there's no harm in it, but I also see that we don't describe `get_keys_function`, should we delete that description?
https://github.com/valkey-io/valkey/pull/2309#discussion_r2625029277 for reference

madolson replied:
I guess there isn't any harm, so that seems ok.

dvkashapov committed:
Apply review suggestions

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
Merge remote-tracking branch 'upstream/unstable' into unstable

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
Update lua and module api to use new VM_ACL check

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

madolson commented on tests/unit/acl-v2.tcl:
```diff
@@ -544,3 +586,876 @@ start_server [list overrides [list "dir" $server_path "aclfile" "userwithselecto
+        $r2 select 0
+    }
+    
+        test {Test MULTI with SELECT to invalid database} {
```
```suggestion
    test {Test MULTI with SELECT to invalid database} {
```

dvkashapov replied:
Thanks!

madolson commented on src/commands/cluster-countkeysinslot.json:
```diff
@@ -8,7 +8,8 @@
         "command_flags": [
-            "STALE"
+            "STALE",
+            "ALL_DBS"
```
Checked the code, this command also only accesses the current db's data, so I think we can remove the ALL_DBS flag.

dvkashapov replied:
Thank you, I should have did the same :(

madolson commented on src/module.c:
There is a function in this file, `commandFlagsFromString`, which should allow modules to attach the ALL_DBs flag.

dvkashapov replied:
Thank you, added attachment and updated for cmd creation.

madolson commented on src/commands/migrate.json:
```diff
@@ -34,6 +34,7 @@
             "SLOW",
             "WRITE"
         ],
+        "get_dbid_args": "migrateDbIdArgs",
```
So, what this command really does is execute a `SELECT <N>; RESTORE key` on the target node not on the current node. So it shouldn't really matter if the current client has access to the permissions to a given db on the current node. @hpatro would also appreciate your thoughts. 

```suggestion
```

dvkashapov replied:
There was an idea that we should fail fast if we don’t have access on current node, I think both opinions make sense and are valid, not sure what’s the best way to handle this, I guess most of the users sync ACLs across nodes.
https://github.com/valkey-io/valkey/pull/2309#issuecomment-3538976222

zuiderkwast replied:
For cluster slot migration, users probably sync ACL across nodes in the cluster, but there are other use cases for MIGRATE. Do tools like RediShake and Riot use MIGRATE?

MIGRATE can also take an AUTH argument with username and password for the other node, which may be different to the auth used on the current node.
Maybe it's safest to not assume anything about ACL on other nodes.

dvkashapov replied:
Yeah, this makes sense, deleted this logic.

hpatro replied:
Didn't realize the transformation. Looks good to be removed.

madolson commented on src/acl.c:
```diff
@@ -937,6 +963,71 @@ static void ACLAddAllowedFirstArg(aclSelector *selector, unsigned long id, const
+        }
+
+        /* Reject out of range values */
+        if (errno == ERANGE || dbid < 0 || dbid >= server.dbnum) {
```
```suggestion
        if (errno == ERANGE || dbid < 0) {
```
I'm wondering if we should be verifying the validity of the db number on the current instance. I can imagine someone has a standardized set of ACLs, and they want to apply it widely even if all servers don't have everything configured. In AWS we decouple the configuration of databases from the ACL configuration, and we would need to add extra code to make sure they are consistent. 

We've also talked about making the number of DBs dynamic as well, so if we do we just need to make sure we check that no users have high DBs.

dvkashapov replied:
I think we should have this check because it would prevent a lot of misuse (and also make it faster to search smaller intset compared to intset that has a lot of `dbid >= server.dbnum`).
For the case of dynamic number of DBs, I would assume server administrator has to keep ACLs in sync with current configuration (at Yandex Cloud we also have that logic decoupled but at any given moment for one node there is only one valid server.dbnum, so it shouldn't be that hard to validate, WDYT?)

hpatro replied:
I agree with @dvkashapov to keep it bounded as we will always have a upper bound.

madolson reviewed (commented):
The code looks really good. I looked more deeply into a few things and have some minor suggestions, other than that it looks good!

dvkashapov committed:
Merge remote-tracking branch 'upstream/unstable' into unstable

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
Apply review suggestions

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
Del test for COUNTKEYSINSLOT

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
Delete get_dbid_args for MIGRATE

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>

dvkashapov committed:
Del dbid >= server.dbnum

Signed-off-by: Daniil Kashapov <daniil.kashapov.ykt@gmail.com>
